Add accept-language support

muffinresearch opened the pull request:
* Add accept-language fallback to lang lookup (this is derived from mozilla/i18n-abide code + tests).
* remove lang query-string support since it's superfluous.
* Make routes with lang (optional) makes it possible to redirect if lang isn't supplied.
* Improve server tests, splits them up so that passing NODE_APP_INSTANCE is done correctly - which required some test fixes since not passing the NODE_APP_INSTANCE meant some of the older tests were wrong.

mstriemer commented on src/amo/routes.js:
```diff
@@ -5,8 +5,7 @@ import App from './containers/App';
 
 export default (
-  <Route path="/" component={App}>
+  <Route path="/(:lang/)" component={App}>
```
This should probably be `path="/(:lang/)firefox/"` but that could wait if you'd like.

mstriemer replied:
I wonder how this will work if you get something like `/firefox/addons/`, would it be smart enough to see that there is no `lang`? Maybe we shouldn't worry about redirecting the lang other than at `/`?

muffinresearch replied:
Yes this does work as long as the rest of the URL matches. See https://github.com/mozilla/addons-frontend/pull/770/files#diff-4f380033f80b73f222c107868068167aR72

mstriemer commented on src/core/i18n/utils.js:
```diff
@@ -73,24 +73,93 @@ export function getLanguage(langOrLocale) {
+ * Parses the HTTP accept-language header and returns a
+ * sorted array of objects. Example object:
+ * {
+ *   lang: 'pl', quality: 0.7
```
I feel like this object could be on one line.

mstriemer commented on src/core/i18n/utils.js:
```diff
@@ -73,24 +73,93 @@ export function getLanguage(langOrLocale) {
+  const langList = rawLangs.map((rawLang) => {
+    const parts = rawLang.split(';');
+    let q = 1;
+    if (parts.length > 1 && parts[1].indexOf('q=') === 0) {
```
Here's an example from [Wikipedia](https://en.wikipedia.org/wiki/Content_negotiation): `Accept-Language: de; q=1.0, en; q=0.5`. I don't think this would parse it.

Maybe we use a URL parsing library to handle this? Example: https://tonicdev.com/5728c8c86ad20d1100700899/578937d636f3d6120002d359

mstriemer commented on src/core/i18n/utils.js:
```diff
@@ -73,24 +73,93 @@ export function getLanguage(langOrLocale) {
+
+/*
+ * Given an accept-language header and a list of currently
+ * supported languages, returns the best match with no normalization.
```
This says "with no normalization" but `isValidLang` is `_validLangs.includes(normalizeLang(lang))` which appears to normalize?

mstriemer replied:
Oh, I guess this would match `en-CA` to `en` and then return `en-CA`? Is that wanted?

mstriemer replied:
Ah, never mind. It would treat `en-ca` as `en-CA` and not match, then pick `en` since it matches without a region.

mstriemer commented on src/core/i18n/utils.js:
```diff
@@ -73,24 +73,93 @@ export function getLanguage(langOrLocale) {
+    userLang = renderProps.params.lang;
+  }
+  // If we don't have a valid userLang set yet try accept-language.
+  if (!isValidLang(userLang) && acceptLanguage) {
```
I think the `!isValidLang()` bit would be clearer with the `renderProps`. I feel we should be able to trust `renderProps` is formatted correctly as well.

```js
export function getFilteredUserLanguage({ renderProps, acceptLanguage } = {}) {
  if (isValidLang(renderProps.params.lang)) {
    return renderProps.params.lang;
  }
  return sanitizeLanguage(getLangFromHeader(acceptLanguage));
}

muffinresearch replied:
I don't know how we can trust renderProps is OK without checking it against the list of allowed values. It feels safer to assume we can't trust it to me.

mstriemer replied:
I just mean we should be able to trust that react-router gives us the same `renderProps` structure. So no need to do `renderProps && renderProps.params && renderProps.params.lang`. I think my code is missing a `sanitizeLanguage()` call in the `if`.

It feels like this function is doing a bit too much since it is finding possible langs, validating them and returning a match. I'm not sure if it's worth the refactor but it would be reasonable to me to make this essentially just do `isValidLang(lang) ? sanitizeLang(lang) : defaultLang`.

muffinresearch replied:
Ah I see now. One problem is that not all apps will have localization.

mstriemer commented on tests/client/core/i18n/test_utils.js:
```diff
@@ -82,25 +82,25 @@ describe('i18n utils', () => {
 
     it('should return a lang if handed a locale', () => {
-      assert.equal(utils.getLanguage('en_US'), 'en-US');
+      assert.equal(utils.sanitizeLanguage('en_US'), 'en-US');
```
Can you add `assert.equal(utils.sanitizeLanguage(undefined), 'en-US');`?

mstriemer commented on tests/client/core/i18n/test_utils.js:
```diff
@@ -156,32 +181,146 @@ describe('i18n utils', () => {
+      assert.deepEqual(utils.parseAcceptLanguage(input), [
+        { lang: 'en', quality: 0.7 },
+        { lang: 'fil', quality: 0.5 },
+      ], JSON.stringify(utils.parseAcceptLanguage(input)));
```
You can use `sinon.format(...)` to get nicer output.

mstriemer commented:
r+wc

muffinresearch commented:
@mstriemer I've added a new commit which removes the index route and redirects in development. This makes it a bit more correct so hitting / will be a 404 which I think is right. I also moved from 301 to 302 for redirection which feels more appropriate.

mstriemer commented on src/core/server/base.js:
```diff
@@ -83,6 +83,11 @@ function baseServer(routes, createStore, { appInstanceName = appName } = {}) {
   }
 
+  if (appInstanceName === 'disco' && isDevelopment) {
+    app.get('/', (req, res) =>
```
react-router has a [`<Redirect />`](https://github.com/reactjs/react-router/blob/master/docs/API.md#redirect) component, can you add that to the router in disco/routes.js when in development instead?

muffinresearch commented:
> react-router has a <Redirect /> component, can you add that to the router in disco/routes.js when in development instead?

That would be nice for removing app specific code from the core apps. What I can't see is what status code you'll get using that.

muffinresearch commented:
I suppose we could say for local-development we don't really care.

mstriemer commented:
Actually react-router won't perform the redirect on the server, so I'm not sure what will happen. Maybe it returns a `200` then redirects in the browser?

muffinresearch commented:
> Actually react-router won't perform the redirect on the server, so I'm not sure what will happen. Maybe it returns a 200 then redirects in the browser?

Yeah - I can't seem to get it to work - presumably because you need to respond to a URL you don't care about in order to be running the JS which will do the location change :/

muffinresearch commented:
I'm going to land this as is, and we can look at improving the redirects stuff in due course.
